### PR TITLE
Remove reliance on pkg_resources for entry points

### PR DESCRIPTION
**Pull request by jaraco**
Ref #3085

### CODE REVIEW

**Comment by jaraco** [comment]
That's weird. Tests are timing out on PyPy (only). The same thing happens on my local machine.

Curiously, if I disable the pytest-xdist plugin, the tests run. So it seems xdist is implicated.

**Comment by jaraco** [comment]
I do see that at least one test passed on pypy for SETUPTOOLS_USE_DISTUTILS=local. I don't get it - there's next to nothing in this change that's related to distutils or pypy.

**Comment by jaraco** [comment]
An earlier test is passing or nearly passing. Maybe the issue is transient.

**Comment by jaraco** [comment]
Issue was apparently transient. Seems tests are getting flakier and flakier. Not good considering it’s now taking hours to work through jobs.

**Comment by abravalheri** [comment]
> Issue was apparently transient. Seems tests are getting flakier and flakier. Not good considering it’s now taking hours to work through jobs.

Yes, I noticed in previous PRs that the combination Windows + PyPy 3.7 is specially flaky. In most of the cases I observed, the error happens on `test_build_meta.py` with `self.pool.submit`. For some reason the combination seems to lead to memory allocation errors (lack of available memory in the worker, maybe? That would make sense with the behaviour you observed with `xdist`). Other time it just takes for ever and ends up cancelled.

These are just a few of the examples I found recently:
- https://github.com/pypa/setuptools/runs/5056604070?check_suite_focus=true#step:5:1583
- https://github.com/pypa/setuptools/runs/5021063326?check_suite_focus=true#step:5:1690

File: changelog.d/3085.change.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Setuptools no longer relies on pkg_resources for entry point handling.
```
Comment (webknjaz) [inline changelog.d/3085.change.rst:1]: (nit)
```suggestion
Setuptools no longer relies on ``pkg_resources`` for entry point handling.
```

File: setuptools/_entry_points.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    return ep.group, ep.name
+
+
+def validate(eps: metadata.EntryPoints):
+    """
+    Ensure entry points are unique by group and name and validate the pattern.
+    """
+    for ep in ensure_unique(eps, key=by_group_and_name):
+        # exercise one of the dynamic properties to trigger validation
+        ep.extras
```
Comment (webknjaz) [inline setuptools/_entry_points.py:41]: Maybe consider reusing the helper?
```suggestion
        ensure_valid(ep)
```

Comment (jaraco) [reply setuptools/_entry_points.py:41]: Thanks for flagging this, as I'd lost track of it.

You're right. I was noticing that `ensure_valid` wasn't used anywhere. Some consolidation is called for. If I was going to invoke `ensure_valid`, I'd also want to remove the preceding comment, which becomes redundant, so I'll figure something out separately.

Comment (jaraco) [reply setuptools/_entry_points.py]: Fixed in f4b0e3ac5.

**Pull request merged**